File: servers/golang/starfish/data_store_test.go
package starfish

import (
	"encoding/json"
	"testing"
)

func TestDataStoreReplace(t *testing.T) {
	ds := NewDataStore()

	entry, err := ds.Apply("replace", "score", "session", "c1", json.RawMessage(`42`), nil)
	if err != nil {
		t.Fatal(err)
	}
	if entry.Version != 1 {
		t.Fatalf("expected version 1, got %d", entry.Version)
	}
	if string(entry.Data) != "42" {
		t.Fatalf("expected 42, got %s", entry.Data)
	}

	// Replace again
	entry, err = ds.Apply("replace", "score", "session", "c1", json.RawMessage(`100`), nil)
	if err != nil {
		t.Fatal(err)
	}
	if entry.Version != 2 {
		t.Fatalf("expected version 2, got %d", entry.Version)
	}
}

func TestDataStoreMerge(t *testing.T) {
	ds := NewDataStore()

	ds.Apply("replace", "config", "session", "c1", json.RawMessage(`{"a":1,"b":2}`), nil)

	entry, err := ds.Apply("merge", "config", "session", "c1", json.RawMessage(`{"b":3,"c":4}`), nil)
	if err != nil {
		t.Fatal(err)
	}

	var result map[string]float64
	json.Unmarshal(entry.Data, &result)
	if result["a"] != 1 || result["b"] != 3 || result["c"] != 4 {
		t.Fatalf("unexpected merge result: %s", entry.Data)
	}
}

func TestDataStoreCounterAdd(t *testing.T) {
	ds := NewDataStore()

	ds.Apply("counter.add", "points", "session", "c1", json.RawMessage(`10`), nil)
	entry, _ := ds.Apply("counter.add", "points", "session", "c1", json.RawMessage(`5`), nil)

	var val float64
	json.Unmarshal(entry.Data, &val)
	if val != 15 {
		t.Fatalf("expected 15, got %f", val)
	}
}

func TestDataStoreSetAddRemove(t *testing.T) {
	ds := NewDataStore()

	ds.Apply("set.add", "tags", "session", "c1", json.RawMessage(`"a"`), nil)
	ds.Apply("set.add", "tags", "session", "c1", json.RawMessage(`"b"`), nil)
	ds.Apply("set.add", "tags", "session", "c1", json.RawMessage(`"a"`), nil) // Duplicate

	entry := ds.Get("tags", "session", "c1")
	var set []string
	json.Unmarshal(entry.Data, &set)
	if len(set) != 2 {
		t.Fatalf("expected 2 items, got %d: %v", len(set), set)
	}

	ds.Apply("set.remove", "tags", "session", "c1", json.RawMessage(`"a"`), nil)
	entry = ds.Get("tags", "session", "c1")
	json.Unmarshal(entry.Data, &set)
	if len(set) != 1 || set[0] != "b" {
		t.Fatalf("expected [b], got %v", set)
	}
}

func TestDataStoreListAddRemove(t *testing.T) {
	ds := NewDataStore()

	ds.Apply("list.add", "log", "session", "c1", json.RawMessage(`"event1"`), nil)
	ds.Apply("list.add", "log", "session", "c1", json.RawMessage(`"event2"`), nil)
	ds.Apply("list.add", "log", "session", "c1", json.RawMessage(`"event1"`), nil) // Duplicates allowed in list

	entry := ds.Get("log", "session", "c1")
	var list []string
	json.Unmarshal(entry.Data, &list)
	if len(list) != 3 {
		t.Fatalf("expected 3 items, got %d", len(list))
	}

	ds.Apply("list.remove", "log", "session", "c1", json.RawMessage(`"event1"`), nil)
	entry = ds.Get("log", "session", "c1")
	json.Unmarshal(entry.Data, &list)
	if len(list) != 1 || list[0] != "event2" {
		t.Fatalf("expected [event2], got %v", list)
	}
}

func TestDataStoreDelete(t *testing.T) {
	ds := NewDataStore()

	ds.Apply("replace", "tmp", "session", "c1", json.RawMessage(`"value"`), nil)
	entry, _ := ds.Apply("delete", "tmp", "session", "c1", nil, nil)

	if entry.Version != 2 {
		t.Fatalf("expected version 2, got %d", entry.Version)
	}

	got := ds.Get("tmp", "session", "c1")
	if got.Version != 0 {
		t.Fatalf("expected version 0 for deleted key, got %d", got.Version)
	}
}

func TestDataStoreOptimisticConcurrency(t *testing.T) {
	ds := NewDataStore()

	ds.Apply("replace", "x", "session", "c1", json.RawMessage(`1`), nil)

	// Correct version
	v := int64(1)
	_, err := ds.Apply("replace", "x", "session", "c1", json.RawMessage(`2`), &v)
	if err != nil {
		t.Fatalf("expected success, got %v", err)
	}

	// Wrong version
	v = int64(1)
	_, err = ds.Apply("replace", "x", "session", "c1", json.RawMessage(`3`), &v)
	if err == nil {
		t.Fatal("expected conflict error")
	}
	conflict, ok := err.(*ConflictError)
	if !ok {
		t.Fatalf("expected ConflictError, got %T", err)
	}
	if conflict.ActualVersion != 2 {
		t.Fatalf("expected actual version 2, got %d", conflict.ActualVersion)
	}
}

func TestDataStoreScopeIsolation(t *testing.T) {
	ds := NewDataStore()

	ds.Apply("replace", "secret", "self", "c1", json.RawMessage(`"mine"`), nil)
	ds.Apply("replace", "secret", "self", "c2", json.RawMessage(`"yours"`), nil)

	e1 := ds.Get("secret", "self", "c1")
	e2 := ds.Get("secret", "self", "c2")

	if string(e1.Data) != `"mine"` {
		t.Fatalf("expected mine, got %s", e1.Data)
	}
	if string(e2.Data) != `"yours"` {
		t.Fatalf("expected yours, got %s", e2.Data)
	}

	// Session scope is separate
	ds.Apply("replace", "secret", "session", "c1", json.RawMessage(`"shared"`), nil)
	es := ds.Get("secret", "session", "c1")
	if string(es.Data) != `"shared"` {
		t.Fatalf("expected shared, got %s", es.Data)
	}
}

func TestDataStoreNewKeyExpectedVersion(t *testing.T) {
	ds := NewDataStore()

	// expectedVersion 0 for a new key should succeed
	v := int64(0)
	_, err := ds.Apply("replace", "new", "session", "c1", json.RawMessage(`"first"`), &v)
	if err != nil {
		t.Fatalf("expected success for expectedVersion 0 on new key, got %v", err)
	}
}

File: servers/typescript/src/data_store.test.ts
import { describe, it, expect } from "vitest";
import { DataStore, ConflictError } from "./data_store.js";

// --- DataStore unit tests ---

describe("DataStore", () => {
  it("replace sets and overwrites values", () => {
    const ds = new DataStore();

    const entry = ds.apply("replace", "score", "session", "c1", 42);
    expect(entry.version).toBe(1);
    expect(entry.data).toBe(42);

    const entry2 = ds.apply("replace", "score", "session", "c1", 100);
    expect(entry2.version).toBe(2);
    expect(entry2.data).toBe(100);
  });

  it("merge shallow-merges objects", () => {
    const ds = new DataStore();

    ds.apply("replace", "config", "session", "c1", { a: 1, b: 2 });
    const entry = ds.apply("merge", "config", "session", "c1", { b: 3, c: 4 });

    expect(entry.data).toEqual({ a: 1, b: 3, c: 4 });
  });

  it("merge on empty key creates object", () => {
    const ds = new DataStore();

    const entry = ds.apply("merge", "config", "session", "c1", { x: 1 });
    expect(entry.data).toEqual({ x: 1 });
  });

  it("merge rejects non-object incoming", () => {
    const ds = new DataStore();

    expect(() => ds.apply("merge", "k", "session", "c1", "string")).toThrow(
      "merge requires an object",
    );
  });

  it("counter.add accumulates numbers", () => {
    const ds = new DataStore();

    ds.apply("counter.add", "points", "session", "c1", 10);
    const entry = ds.apply("counter.add", "points", "session", "c1", 5);

    expect(entry.data).toBe(15);
  });

  it("counter.add starts from zero", () => {
    const ds = new DataStore();

    const entry = ds.apply("counter.add", "points", "session", "c1", 7);
    expect(entry.data).toBe(7);
  });

  it("counter.add rejects non-number", () => {
    const ds = new DataStore();

    expect(() =>
      ds.apply("counter.add", "k", "session", "c1", "bad"),
    ).toThrow("counter.add requires a number");
  });

  it("set.add adds unique items", () => {
    const ds = new DataStore();

    ds.apply("set.add", "tags", "session", "c1", "a");
    ds.apply("set.add", "tags", "session", "c1", "b");
    ds.apply("set.add", "tags", "session", "c1", "a"); // duplicate

    const entry = ds.get("tags", "session", "c1");
    expect(entry.data).toEqual(["a", "b"]);
    expect(entry.version).toBe(3);
  });

  it("set.remove removes matching items", () => {
    const ds = new DataStore();

    ds.apply("set.add", "tags", "session", "c1", "a");
    ds.apply("set.add", "tags", "session", "c1", "b");
    ds.apply("set.remove", "tags", "session", "c1", "a");

    const entry = ds.get("tags", "session", "c1");
    expect(entry.data).toEqual(["b"]);
  });

  it("list.add allows duplicates", () => {
    const ds = new DataStore();

    ds.apply("list.add", "log", "session", "c1", "event1");
    ds.apply("list.add", "log", "session", "c1", "event2");
    ds.apply("list.add", "log", "session", "c1", "event1"); // duplicate allowed

    const entry = ds.get("log", "session", "c1");
    expect(entry.data).toEqual(["event1", "event2", "event1"]);
  });

  it("list.remove removes all occurrences", () => {
    const ds = new DataStore();

    ds.apply("list.add", "log", "session", "c1", "event1");
    ds.apply("list.add", "log", "session", "c1", "event2");
    ds.apply("list.add", "log", "session", "c1", "event1");
    ds.apply("list.remove", "log", "session", "c1", "event1");

    const entry = ds.get("log", "session", "c1");
    expect(entry.data).toEqual(["event2"]);
  });

  it("delete removes key and increments version", () => {
    const ds = new DataStore();

    ds.apply("replace", "tmp", "session", "c1", "value");
    const entry = ds.apply("delete", "tmp", "session", "c1", undefined);

    expect(entry.version).toBe(2);

    const got = ds.get("tmp", "session", "c1");
    expect(got.version).toBe(0);
    expect(got.data).toBeUndefined();
  });

  it("optimistic concurrency succeeds with correct version", () => {
    const ds = new DataStore();

    ds.apply("replace", "x", "session", "c1", 1);
    const entry = ds.apply("replace", "x", "session", "c1", 2, 1);
    expect(entry.version).toBe(2);
    expect(entry.data).toBe(2);
  });

  it("optimistic concurrency fails with wrong version", () => {
    const ds = new DataStore();

    ds.apply("replace", "x", "session", "c1", 1);
    ds.apply("replace", "x", "session", "c1", 2, 1);

    try {
      ds.apply("replace", "x", "session", "c1", 3, 1);
      expect.unreachable("should have thrown");
    } catch (err) {
      expect(err).toBeInstanceOf(ConflictError);
      const conflict = err as ConflictError;
      expect(conflict.actualVersion).toBe(2);
      expect(conflict.currentData).toBe(2);
    }
  });

  it("expectedVersion 0 works for new keys", () => {
    const ds = new DataStore();

    const entry = ds.apply("replace", "new", "session", "c1", "first", 0);
    expect(entry.version).toBe(1);
  });

  it("scope isolation between self clients", () => {
    const ds = new DataStore();

    ds.apply("replace", "secret", "self", "c1", "mine");
    ds.apply("replace", "secret", "self", "c2", "yours");

    expect(ds.get("secret", "self", "c1").data).toBe("mine");
    expect(ds.get("secret", "self", "c2").data).toBe("yours");
  });

  it("scope isolation between self and session", () => {
    const ds = new DataStore();

    ds.apply("replace", "key", "self", "c1", "private");
    ds.apply("replace", "key", "session", "c1", "shared");

    expect(ds.get("key", "self", "c1").data).toBe("private");
    expect(ds.get("key", "session", "c1").data).toBe("shared");
  });

  it("get returns empty entry for missing key", () => {
    const ds = new DataStore();

    const entry = ds.get("missing", "session", "c1");
    expect(entry.version).toBe(0);
    expect(entry.data).toBeUndefined();
  });

  it("rejects invalid operation", () => {
    const ds = new DataStore();

    expect(() => ds.apply("bad.op", "k", "session", "c1", null)).toThrow(
      "invalid operation: bad.op",
    );
  });
});

